crates/red_knot_python_semantic/src/types.rs

```diff
@@ -5650,31 +5725,65 @@ impl<'db> BoolError<'db> {
+                    Severity::Info,
+                    "`__bool__` methods must only have a `self` parameter",
                 );
+                if let Some((func_span, parameter_span)) = not_boolable_type
+                    .member(context.db(), "__bool__")
+                    .into_lookup_result()
+                    .ok()
+                    .and_then(|quals| quals.inner_type().parameter_span(context.db(), None))
+                {
+                    sub.annotate(
+                        Annotation::primary(parameter_span).message("Incorrect parameters"),
+                    );
+                    sub.annotate(Annotation::secondary(func_span).message("Method defined here"));
+                }
+                diag.sub(sub);
             }
             Self::IncorrectReturnType {
                 not_boolable_type,
                 return_type,
             } => {
-                builder.into_diagnostic(format_args!(
-                    "Boolean conversion is unsupported for type `{not_boolable}`; \
-                     the return type of its bool method (`{return_type}`) \
-                     isn't assignable to `bool",
+                let mut diag = builder.into_diagnostic(format_args!(
+                    "Boolean conversion is unsupported for type `{not_boolable}`",
                     not_boolable = not_boolable_type.display(context.db()),
-                    return_type = return_type.display(context.db())
                 ));
+                let mut sub = SubDiagnostic::new(
+                    Severity::Info,
+                    format_args!(
+                        "`{return_type}` is not assignable to `bool`",
+                        return_type = return_type.display(context.db()),
+                    ),
+                );
+                if let Some((func_span, return_type_span)) = not_boolable_type
+                    .member(context.db(), "__bool__")
+                    .into_lookup_result()
+                    .ok()
+                    .and_then(|quals| quals.inner_type().return_type_span(context.db()))
+                {
+                    sub.annotate(
+                        Annotation::primary(return_type_span).message("Incorrect return type"),
+                    );
+                    sub.annotate(Annotation::secondary(func_span).message("Method defined here"));
+                }
+                diag.sub(sub);
             }
             Self::NotCallable { not_boolable_type } => {
-                builder.into_diagnostic(format_args!(
-                    "Boolean conversion is unsupported for type `{}`; \
-                     its `__bool__` method isn't callable",
+                let mut diag = builder.into_diagnostic(format_args!(
+                    "Boolean conversion is unsupported for type `{}`",
                     not_boolable_type.display(context.db())
                 ));
+                let sub = SubDiagnostic::new(Severity::Info, "`__bool__` must be callable");
```

Comment:
I think one fairly simple thing we could do here to improve this diagnostic (short of the below) would be to name the not-callable type we found for `__bool__`. That alone could really help someone to track down the source of the issue.

Reply:
Is that not done in the couple lines above this one? Or am I misunderstanding? That is, I think `not_boolable_type` is the not-callable type, and its name is included in the main diagnostic message.

Or maybe you're saying I should repeat it here? That is, something like:

```
`__bool__` on `{not_boolable_type}` must be callable
```

Reply:
I ended up making that change.